File: parent/143 -- Reorder List/tutorial_mine.py


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Iterative reversal
        curr = head
        prev = None
        
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return prev
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # Time: O(n)
        # Space: O(1)
        # Find midpoint
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Reverse 2nd half
        tail = self.reverseList(slow)

        # loop and insert
        while tail.next:
            nxt = head.next         # temp variable
            head.next = tail
            tail = tail.next        # move tail ptr
            head.next.next = nxt
            head = head.next.next   # move head ptr
        tail.next = None    # set last tail node next ptr to null

File: parent/143 -- Reorder List/tutorial_mine.go
package main

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev, curr *ListNode
	curr = head
	for curr != nil {
		nxt := curr.Next
		curr.Next = prev
		prev = curr
		curr = nxt
	}
	return prev
}
func reorderList(head *ListNode) {
	fast, slow := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	tail := reverseList(slow)
	for tail.Next != nil {
		nxt := head.Next
		head.Next = tail
		tail = tail.Next
		head.Next.Next = nxt
		head = head.Next.Next
	}
	tail.Next = nil
}
